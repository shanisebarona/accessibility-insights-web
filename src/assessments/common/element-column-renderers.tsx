// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { InstanceTableRow } from 'assessments/types/instance-table-data';
import { AssessmentInstanceDetailsColumn } from 'DetailsView/components/assessment-instance-details-column';
import * as React from 'react';

export function onRenderPathColumn(item: InstanceTableRow): JSX.Element {
    let textContent = '';
    if (item.instance.target) {
        textContent = item.instance.target.join(';');
    }

    return (
        <AssessmentInstanceDetailsColumn
            textContent={textContent}
            customClassName="not-applicable"
        />
    );
}

export function onRenderSnippetColumn(item: InstanceTableRow): JSX.Element {
    return (
        <AssessmentInstanceDetailsColumn
            textContent={item.instance.html}
            customClassName="not-applicable"
        />
    );
}
