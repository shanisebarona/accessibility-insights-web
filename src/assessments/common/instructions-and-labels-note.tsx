// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import * as React from 'react';

import * as Markup from '../markup';

export const InstructionsAndLabelsNotes = () => (
    <Markup.Emphasis>
        Note: Both WCAG 2.0 and 2.1 require a widget's visible label and instructions (if present)
        to be programmatically determinable. WCAG 2.1 also requires a widget's visible label and
        instructions (if present) to be included in its accessible name and description.
    </Markup.Emphasis>
);
