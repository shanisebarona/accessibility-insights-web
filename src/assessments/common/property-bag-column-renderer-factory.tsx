// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { InstanceTableRow } from 'assessments/types/instance-table-data';
import { ColumnValueBag } from 'common/types/property-bag/column-value-bag';
import { PropertyBagColumnRendererConfig } from 'common/types/property-bag/property-bag-column-renderer-config';
import { propertyBagColumnRenderer } from './property-bag-column-renderer';

export class PropertyBagColumnRendererFactory {
    public static getRenderer<TPropertyBag extends ColumnValueBag>(
        configs: PropertyBagColumnRendererConfig<TPropertyBag>[],
    ): (item: InstanceTableRow<TPropertyBag>) => JSX.Element {
        return item => {
            return propertyBagColumnRenderer(item, configs);
        };
    }
}
