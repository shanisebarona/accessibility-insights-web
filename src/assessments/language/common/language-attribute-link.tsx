// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NewTabLink } from 'common/components/new-tab-link';
import * as React from 'react';

export const LanguageAttributeLink = () => (
    <NewTabLink href="http://www.iana.org/assignments/language-subtag-registry/language-subtag-registry">
        language attribute
    </NewTabLink>
);
