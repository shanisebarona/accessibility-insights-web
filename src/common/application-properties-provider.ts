// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { ToolData } from './types/store-data/unified-data-interface';

export const createToolData = (
    scanEngineName: string,
    scanEngineVersion: string,
    toolName: string,
    toolVersion?: string,
    environmentName?: string,
    resolution?: string,
): ToolData => {
    return {
        scanEngineProperties: {
            name: scanEngineName,
            version: scanEngineVersion,
        },
        applicationProperties: {
            name: toolName,
            version: toolVersion,
            environmentName: environmentName,
            resolution: resolution,
        },
    };
};
