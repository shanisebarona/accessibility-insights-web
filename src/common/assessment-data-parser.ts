// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { VersionedAssessmentData } from 'common/types/versioned-assessment-data';

export class AssessmentDataParser {
    public parseAssessmentData = (assessmentData: string): VersionedAssessmentData => {
        const loadedAssessmentData: VersionedAssessmentData = JSON.parse(assessmentData);
        return loadedAssessmentData;
    };
}
