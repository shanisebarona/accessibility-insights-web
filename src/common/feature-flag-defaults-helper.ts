// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { FeatureFlagDetail } from 'common/feature-flags';
import { forEach } from 'lodash';

import { FeatureFlagStoreData } from './types/store-data/feature-flag-store-data';

export class FeatureFlagDefaultsHelper {
    public constructor(private readonly getFeatureFlagDetails: () => FeatureFlagDetail[]) {}

    public getDefaultFeatureFlagValues(): FeatureFlagStoreData {
        const details: FeatureFlagDetail[] = this.getFeatureFlagDetails();
        const values: FeatureFlagStoreData = {};
        forEach(details, detail => {
            values[detail.id] = detail.defaultValue;
        });
        return values;
    }

    public getForceDefaultFlags(): string[] {
        const details: FeatureFlagDetail[] = this.getFeatureFlagDetails();
        const forceDefaultFlags: string[] = [];
        forEach(details, detail => {
            if (detail.forceDefault) {
                forceDefaultFlags.push(detail.id);
            }
        });
        return forceDefaultFlags;
    }
}
