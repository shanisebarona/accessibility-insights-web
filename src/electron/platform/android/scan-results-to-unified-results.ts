// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { UUIDGenerator } from 'common/uid-generator';
import { DictionaryStringTo } from 'types/common-types';
import { AndroidScanResults, RuleResultsData, ViewElementData } from './android-scan-results';
import { RuleInformation } from './rule-information';
import { RuleInformationProviderType } from './rule-information-provider-type';

export type ConvertScanResultsToUnifiedResultsDelegate = (
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
) => UnifiedResult[];

export function convertScanResultsToUnifiedResults(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
): UnifiedResult[] {
    if (!scanResults || !scanResults.ruleResults) {
        return [];
    }

    return createUnifiedResultsFromScanResults(scanResults, ruleInformationProvider, uuidGenerator);
}

function createUnifiedResultsFromScanResults(
    scanResults: AndroidScanResults,
    ruleInformationProvider: RuleInformationProviderType,
    uuidGenerator: UUIDGenerator,
): UnifiedResult[] {
    const viewElementLookup: DictionaryStringTo<ViewElementData> =
        createViewElementLookup(scanResults);
    const unifiedResults: UnifiedResult[] = [];

    for (const ruleResult of scanResults.ruleResults) {
        const ruleInformation: RuleInformation = ruleInformationProvider.getRuleInformation(
            ruleResult.ruleId,
        );

        if (ruleInformation) {
            unifiedResults.push(
                createUnifiedResult(ruleInformation, ruleResult, viewElementLookup, uuidGenerator),
            );
        }
    }

    return unifiedResults;
}

function createViewElementLookup(
    scanResults: AndroidScanResults,
): DictionaryStringTo<ViewElementData> {
    const viewElementLookup = {};

    addViewElementAndChildren(viewElementLookup, scanResults.viewElementTree);

    return viewElementLookup;
}

function addViewElementAndChildren(
    viewElementLookup: DictionaryStringTo<ViewElementData>,
    element: ViewElementData | null,
): void {
    if (element) {
        viewElementLookup[element.axeViewId] = element;
        if (element.children) {
            for (const child of element.children) {
                addViewElementAndChildren(viewElementLookup, child);
            }
        }
    }
}

function createUnifiedResult(
    ruleInformation: RuleInformation,
    ruleResult: RuleResultsData,
    viewElementLookup: DictionaryStringTo<ViewElementData>,
    uuidGenerator: UUIDGenerator,
): UnifiedResult {
    const viewElement = viewElementLookup[ruleResult.axeViewId];
    return {
        uid: uuidGenerator(),
        ruleId: ruleInformation.ruleId,
        status: ruleInformation.getResultStatus(ruleResult),
        descriptors: {
            className: viewElement?.className,
            boundingRectangle: viewElement?.boundsInScreen,
            contentDescription: viewElement?.contentDescription,
            text: viewElement?.text,
        },
        identifiers: {
            identifier: viewElement?.className,
            conciseName: viewElement?.className,
        },
        resolution: ruleInformation.getUnifiedResolution(ruleResult),
    };
}
