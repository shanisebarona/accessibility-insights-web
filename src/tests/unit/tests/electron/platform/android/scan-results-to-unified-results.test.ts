// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { UnifiedResult } from 'common/types/store-data/unified-data-interface';
import { generateUID } from 'common/uid-generator';
import {
    AndroidScanResults,
    RuleResultsData,
} from 'electron/platform/android/android-scan-results';
import { RuleInformation } from 'electron/platform/android/rule-information';
import { RuleInformationProviderType } from 'electron/platform/android/rule-information-provider-type';
import { convertScanResultsToUnifiedResults } from 'electron/platform/android/scan-results-to-unified-results';
import { IMock, It, Mock, MockBehavior, Times } from 'typemoq';
import {
    buildRuleInformation,
    buildRuleResultObject,
    buildScanResultsObject,
    buildViewElement,
} from './scan-results-helpers';

describe('ScanResultsToUnifiedResults', () => {
    let generateGuidMock: IMock<() => string>;
    let ruleInformationProviderMock: IMock<RuleInformationProviderType>;

    const ruleId1: string = 'My Rule #1';
    const ruleId2: string = 'My Rule #2';
    const ruleId3: string = 'My Rule #3';
    const ruleId4: string = 'My Rule #4';

    beforeEach(() => {
        const guidStub = 'gguid-mock-stub';
        generateGuidMock = Mock.ofInstance(generateUID, MockBehavior.Strict);
        generateGuidMock.setup(ggm => ggm()).returns(() => guidStub);

        const ruleInformation1: RuleInformation = buildRuleInformation(ruleId1);
        const ruleInformation2: RuleInformation = buildRuleInformation(ruleId2);
        const ruleInformation3: RuleInformation = buildRuleInformation(ruleId3);
        const ruleInformation4: RuleInformation = buildRuleInformation(ruleId4, 'rule-link-4', []);

        ruleInformationProviderMock = Mock.ofType<RuleInformationProviderType>();
        ruleInformationProviderMock
            .setup(x => x.getRuleInformation(ruleId1))
            .returns(() => ruleInformation1);
        ruleInformationProviderMock
            .setup(x => x.getRuleInformation(ruleId2))
            .returns(() => ruleInformation2);
        ruleInformationProviderMock
            .setup(x => x.getRuleInformation(ruleId3))
            .returns(() => ruleInformation3);
        ruleInformationProviderMock
            .setup(x => x.getRuleInformation(ruleId4))
            .returns(() => ruleInformation4);
    });

    function verifyMockCounts(
        expectedRule1Count: number,
        expectedRule2Count: number,
        expectedRule3Count: number,
        expectedRule4Count: number,
        expectedOtherCount: number,
    ): void {
        const totalCalls: number =
            expectedRule1Count +
            expectedRule2Count +
            expectedRule3Count +
            expectedRule4Count +
            expectedOtherCount;

        ruleInformationProviderMock.verify(
            x => x.getRuleInformation(ruleId1),
            Times.exactly(expectedRule1Count),
        );
        ruleInformationProviderMock.verify(
            x => x.getRuleInformation(ruleId2),
            Times.exactly(expectedRule2Count),
        );
        ruleInformationProviderMock.verify(
            x => x.getRuleInformation(ruleId3),
            Times.exactly(expectedRule3Count),
        );
        ruleInformationProviderMock.verify(
            x => x.getRuleInformation(ruleId4),
            Times.exactly(expectedRule4Count),
        );
        ruleInformationProviderMock.verify(
            x => x.getRuleInformation(It.isAnyString()),
            Times.exactly(totalCalls),
        );
    }

    test('Null ScanResults input returns empty output', () => {
        const results: UnifiedResult[] = convertScanResultsToUnifiedResults(
            null,
            ruleInformationProviderMock.object,
            null,
        );
        expect(results).toMatchSnapshot();
        verifyMockCounts(0, 0, 0, 0, 0);
    });

    test('ScanResults with no RuleResults returns empty output', () => {
        const scanResults: AndroidScanResults = buildScanResultsObject();
        const results: UnifiedResult[] = convertScanResultsToUnifiedResults(
            scanResults,
            ruleInformationProviderMock.object,
            null,
        );
        expect(results).toMatchSnapshot();
        verifyMockCounts(0, 0, 0, 0, 0);
    });

    test('ScanResults with passes, failures, view IDs, and excluded results', () => {
        const id1: string = 'id1';
        const id2: string = 'id2';
        const id3: string = 'id3';
        const id4: string = 'id4';
        const ruleResults: RuleResultsData[] = [
            buildRuleResultObject(ruleId1, 'FAIL', id1),
            buildRuleResultObject('unsupported Rule #1', 'PASS', id2),
            buildRuleResultObject(ruleId2, 'FAIL', id3),
            buildRuleResultObject(ruleId2, 'PASS', id4),
            buildRuleResultObject('unsupported Rule #2', 'FAIL', id1),
            buildRuleResultObject(ruleId4, 'PASS', id2),
            buildRuleResultObject(ruleId3, 'PASS', id2),
            buildRuleResultObject(ruleId2, 'UNKNOWN', 'does not exist'), // Force "unknown" cases
        ];

        const viewElementTree = buildViewElement(
            id1,
            { top: 0, left: 10, bottom: 800, right: 600 },
            'myClass1',
            'myDescription1',
            'myText1',
            [
                buildViewElement(
                    id2,
                    { top: 10, left: 20, right: 35, bottom: 50 },
                    'myClass2',
                    null,
                    null,
                    [buildViewElement(id3, null, 'myClass3', null, null, null)],
                ),
                buildViewElement(
                    id4,
                    { top: 50, left: 10, right: 500, bottom: 300 },
                    'myClass4',
                    'myDescription4',
                    'myText4',
                    null,
                ),
            ],
        );

        const scanResults: AndroidScanResults = buildScanResultsObject(
            'Some device',
            'Some app',
            ruleResults,
            viewElementTree,
        );
        const results: UnifiedResult[] = convertScanResultsToUnifiedResults(
            scanResults,
            ruleInformationProviderMock.object,
            generateGuidMock.object,
        );
        expect(results).toMatchSnapshot();
        verifyMockCounts(1, 3, 1, 1, 2);
    });
});
