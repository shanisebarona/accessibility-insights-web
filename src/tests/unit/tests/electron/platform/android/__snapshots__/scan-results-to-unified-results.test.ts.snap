// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScanResultsToUnifiedResults Null ScanResults input returns empty output 1`] = `Array []`;

exports[`ScanResultsToUnifiedResults ScanResults with no RuleResults returns empty output 1`] = `Array []`;

exports[`ScanResultsToUnifiedResults ScanResults with passes, failures, view IDs, and excluded results 1`] = `
Array [
  Object {
    "descriptors": Object {
      "boundingRectangle": Object {
        "bottom": 800,
        "left": 10,
        "right": 600,
        "top": 0,
      },
      "className": "myClass1",
      "contentDescription": "myDescription1",
      "text": "myText1",
    },
    "identifiers": Object {
      "conciseName": "myClass1",
      "identifier": "myClass1",
    },
    "resolution": Object {
      "howToFixSummary": "How to fix My Rule #1",
    },
    "ruleId": "My Rule #1",
    "status": "fail",
    "uid": "gguid-mock-stub",
  },
  Object {
    "descriptors": Object {
      "boundingRectangle": null,
      "className": "myClass3",
      "contentDescription": null,
      "text": null,
    },
    "identifiers": Object {
      "conciseName": "myClass3",
      "identifier": "myClass3",
    },
    "resolution": Object {
      "howToFixSummary": "How to fix My Rule #2",
    },
    "ruleId": "My Rule #2",
    "status": "fail",
    "uid": "gguid-mock-stub",
  },
  Object {
    "descriptors": Object {
      "boundingRectangle": Object {
        "bottom": 300,
        "left": 10,
        "right": 500,
        "top": 50,
      },
      "className": "myClass4",
      "contentDescription": "myDescription4",
      "text": "myText4",
    },
    "identifiers": Object {
      "conciseName": "myClass4",
      "identifier": "myClass4",
    },
    "resolution": Object {
      "howToFixSummary": "How to fix My Rule #2",
    },
    "ruleId": "My Rule #2",
    "status": "pass",
    "uid": "gguid-mock-stub",
  },
  Object {
    "descriptors": Object {
      "boundingRectangle": Object {
        "bottom": 50,
        "left": 20,
        "right": 35,
        "top": 10,
      },
      "className": "myClass2",
      "contentDescription": null,
      "text": null,
    },
    "identifiers": Object {
      "conciseName": "myClass2",
      "identifier": "myClass2",
    },
    "resolution": Object {
      "howToFixSummary": "How to fix My Rule #4",
    },
    "ruleId": "My Rule #4",
    "status": "pass",
    "uid": "gguid-mock-stub",
  },
  Object {
    "descriptors": Object {
      "boundingRectangle": Object {
        "bottom": 50,
        "left": 20,
        "right": 35,
        "top": 10,
      },
      "className": "myClass2",
      "contentDescription": null,
      "text": null,
    },
    "identifiers": Object {
      "conciseName": "myClass2",
      "identifier": "myClass2",
    },
    "resolution": Object {
      "howToFixSummary": "How to fix My Rule #3",
    },
    "ruleId": "My Rule #3",
    "status": "pass",
    "uid": "gguid-mock-stub",
  },
  Object {
    "descriptors": Object {
      "boundingRectangle": undefined,
      "className": undefined,
      "contentDescription": undefined,
      "text": undefined,
    },
    "identifiers": Object {
      "conciseName": undefined,
      "identifier": undefined,
    },
    "resolution": Object {
      "howToFixSummary": "How to fix My Rule #2",
    },
    "ruleId": "My Rule #2",
    "status": "pass",
    "uid": "gguid-mock-stub",
  },
]
`;
